src/cdk/a11y/input-modality/input-modality-detector.ts

```diff
@@ -0,0 +1,187 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Inject, Injectable, InjectionToken, OnDestroy, Optional} from '@angular/core';
+import {normalizePassiveListenerOptions, Platform} from '@angular/cdk/platform';
+import {DOCUMENT} from '@angular/common';
+import {Observable, ReplaySubject} from 'rxjs';
+import {distinctUntilChanged} from 'rxjs/operators';
+import {
+  isFakeMousedownFromScreenReader,
+  isFakeTouchstartFromScreenReader,
+} from '../fake-event-detection';
+
+/**
+ * The input modalities detected by this service. Null is used if the input modality is unknown.
+ */
+export type InputModality = 'keyboard' | 'mouse' | 'touch' | null;
+
+/** Options to configure the behavior of the InputModalityDetector. */
+export interface InputModalityDetectorOptions {
+  /** Keys to ignore when detecting keyboard input modality. */
+  ignoreKeys?: string[];
+}
+
+/**
+ * Injectable options for the InputModalityDetector. These are shallowly merged with the default
+ * options.
+ */
+export const INPUT_MODALITY_DETECTOR_OPTIONS =
+  new InjectionToken<InputModalityDetectorOptions>('cdk-input-modality-detector-options');
+
+/**
+ * Default options for the InputModalityDetector.
+ *
+ * Modifier keys are ignored by default (i.e. when pressed won't cause the service to detect
+ * keyboard input modality) for two reasons:
+ *
+ * 1. Modifier keys are commonly used with mouse to perform actions such as 'right click' or 'open
+ *    in new tab', and are thus less representative of actual keyboard interaction.
+ * 2. VoiceOver triggers some keyboard events when linearly navigating with Control + Option (but
+ *    confusingly not with Caps Lock). Thus, to have parity with other screen readers, we ignore
+ *    these keys so as to not update the input modality.
+ */
+export const INPUT_MODALITY_DETECTOR_DEFAULT_OPTIONS: InputModalityDetectorOptions = {
+  ignoreKeys: ['Alt', 'Control', 'Meta', 'Shift'],
```

Comment:
These are all modifier keys so shouldn't we be looking at `KeyboardEvent.altKey` etc.? We have some helpers for dealing with modifiers in `@angular/cdk/keycodes` already.

Reply:
I considered that, but I wanted to give users the ability to configure which keys they'd like to ignore/not ignore, and it seemed easiest to support that via a single `ignoreKeys` API.

Reply:
It could have a `modifiers` array and `ignoreKeys`. If I remember correctly, the keydown will dispatch `Alt` if you just press the `Alt` key once, but if you pressed `Alt` + `DOWN_ARROW`, you'd get a `DOWN_ARROW` event with the `Alt` modifier.

Reply:
From my testing, if I press `Alt` + `Down Arrow`, I get both an `Alt` and a `Down Arrow` event, so this should work fine (I tested with a number of modifier keys and on both Mac and Windows). Changing the API to both `modifiers` and `ignoreKeys` increases its surface error, and it's not clear to me what the benefit is (do we think it'll be clearer to users?).